mlir_graphblas/src/lib/GraphBLAS/GraphBLASOptimizePass.cpp

```diff
@@ -86,30 +86,61 @@ class FuseMatrixSelectRewrite : public OpRewritePattern<graphblas::MatrixSelectO
   };
 };
 
-class FuseMatrixMultiplyReduceRewrite : public OpRewritePattern<graphblas::MatrixReduceToScalarOp>
+class FuseMatrixMultiplyReduceRewrite : public OpRewritePattern<graphblas::MatrixReduceToScalarGenericOp>
 {
 public:
-  using OpRewritePattern<graphblas::MatrixReduceToScalarOp>::OpRewritePattern;
-  LogicalResult matchAndRewrite(graphblas::MatrixReduceToScalarOp op, PatternRewriter &rewriter) const
+  using OpRewritePattern<graphblas::MatrixReduceToScalarGenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(graphblas::MatrixReduceToScalarGenericOp op, PatternRewriter &rewriter) const
   {
     Value input = op.input();
-    graphblas::MatrixMultiplyOp predecessor = input.getDefiningOp<graphblas::MatrixMultiplyOp>();
+    graphblas::MatrixMultiplyGenericOp predecessor = input.getDefiningOp<graphblas::MatrixMultiplyGenericOp>();
     if (predecessor != nullptr && predecessor->hasOneUse()) {
       Location loc = op->getLoc();
 
-      // Build new MatrixMultiplyReduce op with the operands and arguments of the multiply,
+      // Build new MatrixMultiplyReduceToScalarGeneric op with the operands and regions of the multiply,
       // then add in the aggregator from the reduce
       ValueRange operands = predecessor.getOperands();
       NamedAttrList attributes = predecessor->getAttrs();
+      RegionRange multiplyExtensions = predecessor.extensions();
+      unsigned newRegions = multiplyExtensions.size();
```

Comment:
Can we add a check here for the rank of predecessor.a() and predecessor.b()? If either has rank 1, this fusion should fail. The fused version doesn't support mxv, vxm, or vxv (and likely shouldn't because the potential savings would be minimal).

Reply:
I've added a check  (Although can't test this now because matrix_reduce_to_scalar currently only works for rank 2, and the result of mxv, vxm, and vxv all have rank < 2)